File: 1101-1200/1181-Before and After Puzzle/1181-Before and After Puzzle.py
import collections
class Solution:
    def beforeAndAfterPuzzles(self, phrases: List[str]) -> List[str]:
        first = collections.defaultdict(list)
        last = collections.defaultdict(list) 
        result = set()
        for p in phrases:
            index1 = p.rfind(' ')
            la = p[index1 + 1:]
            result.update([p + w for w in first[la]])            
            index2 = p.find(' ')
            fi = p[:index2] if index2 != -1 else p
            result.update([w + p for w in last[fi]])
            
            if index1 != -1:
                last[la].append(p[:index1 + 1])
            else:
                last[la].append('')
            if index2 != -1:
                first[fi].append(p[index2:])
            else:
                first[fi].append('')

        return sorted(result)

File: 1101-1200/1181-Before and After Puzzle/1181-Before and After Puzzle.cpp
class Solution {
public:
    vector<string> beforeAndAfterPuzzles(vector<string>& phrases) {
        unordered_map<string, vector<string>> first, last;
        set<string> result;
        for (string p : phrases) {
            auto index1 = p.rfind(' ');
            string la = (index1 != string::npos) ? p.substr(index1 + 1) : p;
            for (string w : first[la]) {
                result.insert(p + w);
            }
            auto index2 = p.find(' ');
            string fi = (index2 != string::npos) ? p.substr(0, index2) : p;
            for (string w : last[fi]) {
                result.insert(w + p);
            }
            
            if (index1 != string::npos) {
                last[la].push_back(p.substr(0, index1 + 1));
            }
            else {
                last[la].push_back("");
            }
            if (index2 != string::npos) {
                first[fi].push_back(p.substr(index2));
            }
            else {
                first[fi].push_back("");
            }
        }
        return vector<string>(result.begin(), result.end());
    }
};
